core/src/main/java/com/yahoo/oak/FirstLevelHashArray.java

```diff
@@ -17,56 +17,76 @@
 
         // the key hash (int) separation between MSB for first level ans LSB for second level,
         // to be used by hash array and all chunks, until resize
         this.hashIndexCodec =
             new UnionCodec(UnionCodec.AUTO_CALCULATE_BIT_SIZE, // the size of the first, as these are LSBs
-                msbForFirstLevelHash, Integer.SIZE); // the second (MSB) will be msbForFirstLevelHash
+                msbForFirstLevelHash, // the second (MSB) will be msbForFirstLevelHash
+                1, // and disregard the first bit as it is the sign bit and always zero
+                Integer.SIZE);
 
         // the size of the hash array is defined by number of MSBs to be used from key hash
         int arraySize = (int) (Math.pow(2, msbForFirstLevelHash));
         this.chunks = new AtomicReferenceArray<>(arraySize);
         this.msbForFirstLevelHash = msbForFirstLevelHash;
 
+        int currentSameRefer = multipleReferenceNum;
+        HashChunk<K, V> c = null;
+        UnionCodec hashIndexCodecForChunk =
+            (lsbForSecondLevelHash == InternalOakHash.USE_DEFAULT_FIRST_TO_SECOND_BITS_PARTITION)
+                ? this.hashIndexCodec :
+                new UnionCodec(lsbForSecondLevelHash, // the size of the first, as these are LSBs
+                    UnionCodec.AUTO_CALCULATE_BIT_SIZE, Integer.SIZE); // the second (MSB) will be auto-calculated
+        int chunkSize = calculateChunkSize(lsbForSecondLevelHash);
         // initiate chunks
-        for (int i = 0; i < msbForFirstLevelHash; i++) {
-            HashChunk<K, V> c =
-                new HashChunk<>(calculateChunkSize(), externalSize, vMM, kMM, comparator,
-                    keySerializer, valueSerializer, hashIndexCodec);
+        System.out.println("Going to allocate " + chunks.length() + " chunks, of size " + chunkSize);
+        for (int i = 0; i < chunks.length(); i++) {
+            if (currentSameRefer == multipleReferenceNum) {
+                c = new HashChunk<>(chunkSize, externalSize, vMM, kMM, comparator,
+                    keySerializer, valueSerializer, hashIndexCodecForChunk);
+            }
+            System.out.print(i + " ");
             this.chunks.lazySet(i, c);
+            currentSameRefer--;
+            if (currentSameRefer == 0) {
+                currentSameRefer = multipleReferenceNum;
+            }
         }
-
     }
 
-    private int calculateChunkSize() {
+    private int calculateChunkSize(int inputLsbForSecondLevel) {
         // least significant bits remaining
-        int lsbForSecondLevel = Integer.SIZE - msbForFirstLevelHash;
+        int lsbForSecondLevel =
+            (inputLsbForSecondLevel == InternalOakHash.USE_DEFAULT_FIRST_TO_SECOND_BITS_PARTITION)
+                ? Integer.SIZE - msbForFirstLevelHash : inputLsbForSecondLevel;
         // size of HashChunk should be 2^lsbForSecondLevel
         return (int) (Math.pow(2, lsbForSecondLevel) + 1); // +1 to round up
```

Comment:
Note that the result of this calculation is expected to be round, so you mostly just adding a redundant one.
You can use `Math.ceil()` instead.
```suggestion
        return (int) Math.ceil(Math.pow(2, lsbForSecondLevel));
```

Reply:
fixed